MaxText/save_xaot.py

```diff
@@ -0,0 +1,123 @@
+"""
+ Copyright 2023 Google LLC
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ """
+
+""" 
+Save a Cross Ahead of Time Compiled (XAOT) version of train.py's train step
+Generates shaped versions of state and data without ever constructing them, so its possible
+to compile with target hardware (e.g. a large cluster), without using the hardware.
+"""
+
+import max_utils
+import pyconfig
+import jax
+import numpy as np
+from typing import Sequence
+from jax.experimental.topologies import get_topology_desc
+from absl import app
+from jax.sharding import Mesh
+from jax.experimental.serialize_executable import serialize, deserialize_and_load
+import pickle
+
+import train
+
+# Ideally this map exists either in XAOT or we can programmatically call it from elsewhere
+# so the chip layout is consistent across cloud technologies
+def get_topology_mesh(config):
```

Comment:
this should be unified with a vending from XPK. For now, just allow two files that mention the other file and asks for human syncs.

Reply:
Added a hardware_map.py file with comments on it and on the corresponding map in xpk/xpk.py pairing them